File: scripts/coverage.py
#!/usr/bin/env python3
"""KernelEngine Coverage.

Builds + runs the test suite under Clang source-based coverage instrumentation
(native: -fprofile-instr-generate / .profraw / llvm-profdata / llvm-cov;
managed: dotnet test + coverlet), then merges both worlds into a single
ReportGenerator output and prints a structured summary.

Usage:
    python scripts/coverage.py            # full pipeline
    python scripts/coverage.py clean      # wipe build/coverage/
    python scripts/coverage.py report     # re-emit summary from cached data

Everything lives under build/coverage/. The script never writes outside it.

Native coverage scope (post-CMake): the engine build is now root build.zig +
per-plugin build.zig, and almost all engine logic is Zig source, which Zig's
own linker compiles/links directly — it does not go through Clang, so the
old scheme of instrumenting every C/C++ translation unit under src/ no
longer applies (there is no `src/c/kernel/` or `src/cpp/` left to instrument;
both were fully migrated). Verified attempting it anyway on a Zig-linked
shared library fails: Zig's linker rejects Clang's profiling-runtime
relocations ("fatal linker error: unhandled relocation type R_X86_64_PC64
... __llvm_prf_data"). The two GTest suites are the one exception — their
build.zig shells out to the SYSTEM clang++ directly for both compile and
link (an ABI workaround already in place, see tests/c/kernel/build.zig), so
`-Dcoverage=true` there produces real, valid instrumented binaries. That
still surfaces meaningful native signal beyond just the test files
themselves: contract headers with inline definitions get instrumented too,
since they're `#include`d straight into the test translation units.
"""

from __future__ import annotations

import argparse
import datetime
import os
import platform
import shutil
import subprocess
import sys
import xml.etree.ElementTree as ET
from pathlib import Path

# ── Paths ────────────────────────────────────────────────────────────────────

BASE_DIR        = Path(__file__).resolve().parent.parent
COVERAGE_DIR    = BASE_DIR / "build" / "coverage"
ZIG_OUT_DIR     = COVERAGE_DIR / "zig-out"
PROFRAWS_DIR    = COVERAGE_DIR / "profraws"
MANAGED_DIR     = COVERAGE_DIR / "managed"
REPORT_DIR      = COVERAGE_DIR / "report"
MERGED_PROFDATA = COVERAGE_DIR / "merged.profdata"
NATIVE_LCOV     = COVERAGE_DIR / "native.lcov"
SUMMARY_TXT     = COVERAGE_DIR / "summary.txt"

IS_WINDOWS = platform.system().lower() == "windows"
NATIVE_TEST_BINARIES = ["test_ke_kernel", "test_integration_cpp"]

# ── Layer classification ─────────────────────────────────────────────────────
# L3 (auto-generated C# bindings under */Native/Generated/) is excluded from
# analysis up front. "Native" is no longer split into kernel/plugin layers:
# see the module docstring for why only the GTest suites' own translation
# units (plus any contract headers they #include) are instrumented today.

def _norm(p: str) -> str:
    return p.replace("\\", "/")

LAYERS = [
    ("L2", "Native (GTest suites + included headers)",
        lambda p: _norm(p).startswith("tests/") or _norm(p).startswith("src/c/")),
    ("L4", "C# framework",  lambda p: _norm(p).startswith("src/csharp/") and "/Generated/" not in _norm(p)),
]

GAP_THRESHOLD_PCT   = 30.0
GAP_THRESHOLD_LINES = 50

# ── Subprocess helpers ───────────────────────────────────────────────────────

def run(cmd, *, check: bool = True, env: dict | None = None,
        capture: bool = False, cwd: Path | None = None) -> subprocess.CompletedProcess:
    # Long argv (e.g. thousands of profraw paths) is uninteresting noise.
    # Show only the program name and a count of the rest.
    if len(cmd) > 6:
        print(f"  $ {cmd[0]} ... ({len(cmd) - 1} args)")
    else:
        print(f"  $ {' '.join(str(a) for a in cmd)}")
    full_env = os.environ.copy()
    if env: full_env.update(env)
    return subprocess.run(
        [str(a) for a in cmd],
        cwd=str(cwd) if cwd else None,
        env=full_env, check=check, text=True,
        capture_output=capture,
    )

def banner(text: str) -> None:
    print()
    print("-" * 70)
    print(f"  {text}")
    print("-" * 70)

# ── Phase 1: build ────────────────────────────────────────────────────────────

def configure_and_build() -> None:
    banner("[1/4] Build (root build.zig, -Dcoverage=true for the GTest suites)")
    vcpkg_root = os.environ.get("VCPKG_ROOT")
    args = ["zig", "build", "--prefix", str(ZIG_OUT_DIR), "-Dcoverage=true"]
    if vcpkg_root:
        args.append(f"-Dvcpkg-root={vcpkg_root}")
    run(args, cwd=BASE_DIR)

# ── Phase 2: native tests + profile merge ────────────────────────────────────

def run_native_tests() -> list[Path]:
    banner("[2/4] Run native tests (instrumented GTest suites)")
    if PROFRAWS_DIR.exists(): shutil.rmtree(PROFRAWS_DIR)
    PROFRAWS_DIR.mkdir(parents=True)
    bin_dir = ZIG_OUT_DIR / "bin"
    binaries: list[Path] = []
    for name in NATIVE_TEST_BINARIES:
        exe = bin_dir / (f"{name}.exe" if IS_WINDOWS else name)
        if not exe.exists():
            print(f"  (missing {exe} - skipping)")
            continue
        binaries.append(exe)
        # %p (PID) + %m (per-binary id) keeps every run's profraw unique.
        env = {"LLVM_PROFILE_FILE": str(PROFRAWS_DIR / f"{name}-%p-%m.profraw")}
        run([str(exe)], env=env, check=False)
    return binaries

def merge_native_profile(binaries: list[Path]) -> bool:
    profraws = list(PROFRAWS_DIR.glob("*.profraw"))
    if not profraws:
        print("  (no .profraw produced - skipping native LCOV export)")
        return False
    print(f"  Merging {len(profraws)} profraw files -> {MERGED_PROFDATA.name}")
    profraw_list = COVERAGE_DIR / "profraws.list"
    profraw_list.write_text("\n".join(str(p) for p in profraws), encoding="utf-8")
    run(["llvm-profdata", "merge", "-sparse",
         f"-o={MERGED_PROFDATA}", "-f", str(profraw_list)], capture=True)
    if not binaries:
        print("  (no instrumented binaries ran - skipping native LCOV export)")
        return False
    print(f"  Exporting LCOV across {len(binaries)} binaries -> {NATIVE_LCOV.name}")
    first, rest = binaries[0], binaries[1:]
    cmd = ["llvm-cov", "export", "-format=lcov",
           f"-instr-profile={MERGED_PROFDATA}", str(first)]
    cmd += [f"-object={p}" for p in rest]
    with NATIVE_LCOV.open("w", encoding="utf-8") as f:
        subprocess.run(cmd, cwd=BASE_DIR, stdout=f, check=True)
    return True

# ── Phase 3: managed tests ───────────────────────────────────────────────────

def run_managed_tests() -> None:
    banner("[3/4] Run managed (C#) tests")
    if MANAGED_DIR.exists(): shutil.rmtree(MANAGED_DIR)
    MANAGED_DIR.mkdir(parents=True)
    run([
        "dotnet", "test", BASE_DIR / "KernelEngine.slnx",
        "--collect:XPlat Code Coverage",
        "--results-directory", MANAGED_DIR,
        "--nologo", "-v", "m",
        "--",
        "DataCollectionRunSettings.DataCollectors.DataCollector.Configuration.Format=cobertura",
    ], check=False)

# ── Phase 4: unified report ──────────────────────────────────────────────────

def build_report() -> None:
    banner("[4/4] Generate unified report")
    if REPORT_DIR.exists(): shutil.rmtree(REPORT_DIR)
    REPORT_DIR.mkdir(parents=True)
    inputs: list[str] = []
    if NATIVE_LCOV.exists() and NATIVE_LCOV.stat().st_size > 0:
        inputs.append(str(NATIVE_LCOV))
    inputs += [str(p) for p in MANAGED_DIR.rglob("*.xml")]
    if not inputs:
        print("  (no coverage data - nothing to report)")
        return
    run([
        "reportgenerator",
        "-reports:" + ";".join(inputs),
        "-targetdir:" + str(REPORT_DIR),
        "-reporttypes:Cobertura;Html",
        "-sourcedirs:" + str(BASE_DIR),
        # *tests* is no longer excluded: the GTest suites' own .cpp files are
        # now the primary native coverage signal (see module docstring).
        "-filefilters:-*Generated*;-*examples*",
        "-classfilters:-*NativeMethods*;-*NativeAnnotation*;-*NativeTypeName*",
    ], check=False)

# ── Structured summary ───────────────────────────────────────────────────────

def _pct(d: dict) -> float:
    return 100.0 * d["covered"] / d["total"] if d["total"] else 0.0

def _module_for(layer: str, path: str) -> str:
    """Group a file path into a coarse module bucket for the summary table."""
    parts = _norm(path).split("/")
    if layer == "L2":
        # tests/c/kernel/... -> tests/c/kernel ; src/c/<domain>/... -> src/c/<domain>
        return "/".join(parts[:3]) if len(parts) >= 3 else "/".join(parts)
    if layer == "L4":
        # src/csharp/<project>/... -> <project>
        return parts[2] if len(parts) >= 3 else "(unknown)"
    return "(unknown)"

def _short(module: str) -> str:
    parts = module.split("/")
    return parts[-1] if parts else module

def print_summary() -> None:
    cobertura = REPORT_DIR / "Cobertura.xml"
    if not cobertura.exists():
        print("\n  (no Cobertura.xml - run the pipeline first)\n")
        return

    root = ET.parse(cobertura).getroot()

    per_layer: dict[str, dict] = {
        lid: {"covered": 0, "total": 0, "modules": {}} for lid, _, _ in LAYERS
    }
    native  = {"covered": 0, "total": 0}
    managed = {"covered": 0, "total": 0}

    base_uri = _norm(str(BASE_DIR)).rstrip("/") + "/"

    for clazz in root.iter("class"):
        file_path = _norm(clazz.get("filename", ""))
        if file_path.startswith(base_uri):
            file_path = file_path[len(base_uri):]
        lines = list(clazz.iter("line"))
        if not lines: continue
        covered = sum(1 for ln in lines if int(ln.get("hits", "0")) > 0)
        total   = len(lines)

        for lid, _, predicate in LAYERS:
            if not predicate(file_path): continue
            L = per_layer[lid]
            L["covered"] += covered
            L["total"]   += total
            module = _module_for(lid, file_path)
            m = L["modules"].setdefault(module, {"covered": 0, "total": 0})
            m["covered"] += covered
            m["total"]   += total
            (native if lid == "L2" else managed)["covered"] += covered
            (native if lid == "L2" else managed)["total"]   += total
            break

    total = {
        "covered": native["covered"] + managed["covered"],
        "total":   native["total"]   + managed["total"],
    }

    out: list[str] = []
    def w(s: str = "") -> None: out.append(s)

    now = datetime.datetime.now().strftime("%Y-%m-%d %H:%M")
    w("=" * 70)
    w(f"  KernelEngine Coverage - {now}")
    w("=" * 70)
    w()
    w(f"  TOTAL                                          {_pct(total):5.1f}%   {total['covered']}/{total['total']}")
    w()
    w(f"  Native  (C/C++)                                {_pct(native):5.1f}%   {native['covered']}/{native['total']}")
    w(f"  Managed (C#)                                   {_pct(managed):5.1f}%   {managed['covered']}/{managed['total']}")
    w()
    w("  Per layer:")
    for lid, name, _ in LAYERS:
        L = per_layer[lid]
        w(f"    {lid} - {name:<22} {_pct(L):5.1f}%   {L['covered']}/{L['total']}")
        for module, m in sorted(L["modules"].items()):
            w(f"        {_short(module):<38} {_pct(m):5.1f}%")
    w()
    gaps: list[tuple[str, float, int]] = []
    for lid, _, _ in LAYERS:
        for module, m in per_layer[lid]["modules"].items():
            if m["total"] >= GAP_THRESHOLD_LINES and _pct(m) < GAP_THRESHOLD_PCT:
                gaps.append((module, _pct(m), m["total"]))
    if gaps:
        w(f"  Critical gaps (< {GAP_THRESHOLD_PCT:.0f}% AND > {GAP_THRESHOLD_LINES} lines):")
        for module, p, n in sorted(gaps, key=lambda t: t[1]):
            w(f"    - {_short(module):<38} {p:5.1f}%   ({n} lines)")
        w()
    w(f"  HTML report: {REPORT_DIR / 'index.html'}")
    w("=" * 70)

    text = "\n".join(out)
    SUMMARY_TXT.write_text(text, encoding="utf-8")
    print()
    print(text)

# ── Subcommands ──────────────────────────────────────────────────────────────

def cmd_clean() -> None:
    if COVERAGE_DIR.exists():
        print(f"Removing {COVERAGE_DIR}")
        shutil.rmtree(COVERAGE_DIR)
    else:
        print(f"{COVERAGE_DIR} does not exist.")

def cmd_run() -> None:
    configure_and_build()
    binaries = run_native_tests()
    merge_native_profile(binaries)
    run_managed_tests()
    build_report()
    print_summary()

def cmd_report() -> None:
    print_summary()

def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("action", nargs="?", default="run",
                        choices=["run", "clean", "report"],
                        help="run = full pipeline, clean = wipe build/coverage/, "
                             "report = re-emit summary from cached data")
    args = parser.parse_args()
    {"run": cmd_run, "clean": cmd_clean, "report": cmd_report}[args.action]()

if __name__ == "__main__":
    main()

File: scripts/coverage.cs
#!/usr/bin/env dotnet run

// KernelEngine Coverage.
//
// Builds + runs the test suite under Clang source-based coverage instrumentation
// (native: -fprofile-instr-generate / .profraw / llvm-profdata / llvm-cov;
// managed: dotnet test + coverlet), then merges both worlds into a single
// ReportGenerator output and prints a structured summary.
//
// Usage:
//     dotnet run scripts/coverage.cs            # full pipeline
//     dotnet run scripts/coverage.cs clean      # wipe build/coverage/
//     dotnet run scripts/coverage.cs report     # re-emit summary from cached data
//
// Everything lives under build/coverage/. The script never writes outside it.
//
// Native coverage scope: almost all engine logic is Zig source, which Zig's own
// linker compiles/links directly — it does not go through Clang, so instrumenting
// every C/C++ translation unit under src/ does not apply. The two GTest suites are
// the one exception — their build.zig shells out to the SYSTEM clang++ directly
// for both compile and link (an ABI workaround already in place, see
// tests/c/kernel/build.zig), so -Dcoverage=true there produces real, valid
// instrumented binaries. That still surfaces meaningful native signal beyond just
// the test files themselves: contract headers with inline definitions get
// instrumented too, since they're #include'd straight into the test translation
// units.

using System.Diagnostics;
using System.Globalization;
using System.Runtime.CompilerServices;
using System.Xml.Linq;

// Report formatting (percentages) must not vary with the host's locale.
CultureInfo.CurrentCulture = CultureInfo.InvariantCulture;

var action = args.Length > 0 ? args[0] : "run";
switch (action)
{
    case "run": Coverage.CmdRun(); break;
    case "clean": Coverage.CmdClean(); break;
    case "report": Coverage.CmdReport(); break;
    default:
        Console.Error.WriteLine($"Unknown action '{action}'. Expected: run, clean, report.");
        return 1;
}
return 0;

static class Coverage
{
    static readonly string BaseDir = Path.GetFullPath(Path.Combine(ScriptPaths.Dir, ".."));
    static readonly string CoverageDir = Path.Combine(BaseDir, "build", "coverage");
    static readonly string ZigOutDir = Path.Combine(CoverageDir, "zig-out");
    static readonly string ProfrawsDir = Path.Combine(CoverageDir, "profraws");
    static readonly string ManagedDir = Path.Combine(CoverageDir, "managed");
    static readonly string ReportDir = Path.Combine(CoverageDir, "report");
    static readonly string MergedProfdata = Path.Combine(CoverageDir, "merged.profdata");
    static readonly string NativeLcov = Path.Combine(CoverageDir, "native.lcov");
    static readonly string SummaryTxt = Path.Combine(CoverageDir, "summary.txt");

    static readonly string[] NativeTestBinaries = ["test_ke_kernel", "test_integration_cpp"];

    // L3 (auto-generated C# bindings under */Native/Generated/) is excluded from
    // analysis up front. "Native" is no longer split into kernel/plugin layers:
    // see the file header for why only the GTest suites' own translation units
    // (plus any contract headers they #include) are instrumented today.
    static readonly (string Id, string Name, Func<string, bool> Predicate)[] Layers =
    [
        ("L2", "Native (GTest suites + included headers)",
            p => Norm(p).StartsWith("tests/") || Norm(p).StartsWith("src/c/")),
        ("L4", "C# framework",
            p => Norm(p).StartsWith("src/csharp/") && !Norm(p).Contains("/Generated/")),
    ];

    const double GapThresholdPct = 30.0;
    const int GapThresholdLines = 50;

    static string Norm(string p) => p.Replace('\\', '/');

    // ── Subprocess helpers ───────────────────────────────────────────────────

    static int Run(IReadOnlyList<string> cmd, IReadOnlyDictionary<string, string>? env = null,
        bool check = true, string? cwd = null, string? stdoutFile = null)
    {
        // Long argv (e.g. thousands of profraw paths) is uninteresting noise.
        // Show only the program name and a count of the rest.
        Console.WriteLine(cmd.Count > 6
            ? $"  $ {cmd[0]} ... ({cmd.Count - 1} args)"
            : $"  $ {string.Join(' ', cmd)}");

        using var process = new Process
        {
            StartInfo = new ProcessStartInfo(cmd[0])
            {
                WorkingDirectory = cwd ?? "",
                RedirectStandardOutput = stdoutFile is not null,
            },
        };
        foreach (var a in cmd.Skip(1)) process.StartInfo.ArgumentList.Add(a);
        if (env is not null)
            foreach (var (k, v) in env) process.StartInfo.Environment[k] = v;

        process.Start();
        if (stdoutFile is not null)
        {
            using var writer = new StreamWriter(stdoutFile);
            writer.Write(process.StandardOutput.ReadToEnd());
        }
        process.WaitForExit();

        if (check && process.ExitCode != 0)
            throw new InvalidOperationException($"Command failed ({process.ExitCode}): {string.Join(' ', cmd)}");
        return process.ExitCode;
    }

    static void Banner(string text)
    {
        Console.WriteLine();
        Console.WriteLine(new string('-', 70));
        Console.WriteLine($"  {text}");
        Console.WriteLine(new string('-', 70));
    }

    // ── Phase 1: build ───────────────────────────────────────────────────────

    static void ConfigureAndBuild()
    {
        Banner("[1/4] Build (root build.zig, -Dcoverage=true for the GTest suites)");
        var vcpkgRoot = Environment.GetEnvironmentVariable("VCPKG_ROOT");
        List<string> cmdArgs = ["zig", "build", "--prefix", ZigOutDir, "-Dcoverage=true"];
        if (vcpkgRoot is not null) cmdArgs.Add($"-Dvcpkg-root={vcpkgRoot}");
        Run(cmdArgs, cwd: BaseDir);
    }

    // ── Phase 2: native tests + profile merge ───────────────────────────────

    static List<string> RunNativeTests()
    {
        Banner("[2/4] Run native tests (instrumented GTest suites)");
        if (Directory.Exists(ProfrawsDir)) Directory.Delete(ProfrawsDir, recursive: true);
        Directory.CreateDirectory(ProfrawsDir);
        var binDir = Path.Combine(ZigOutDir, "bin");
        var binaries = new List<string>();
        foreach (var name in NativeTestBinaries)
        {
            var exe = Path.Combine(binDir, OperatingSystem.IsWindows() ? $"{name}.exe" : name);
            if (!File.Exists(exe))
            {
                Console.WriteLine($"  (missing {exe} - skipping)");
                continue;
            }
            binaries.Add(exe);
            // %p (PID) + %m (per-binary id) keeps every run's profraw unique.
            var env = new Dictionary<string, string>
            {
                ["LLVM_PROFILE_FILE"] = Path.Combine(ProfrawsDir, $"{name}-%p-%m.profraw"),
            };
            Run([exe], env: env, check: false);
        }
        return binaries;
    }

    static bool MergeNativeProfile(List<string> binaries)
    {
        var profraws = Directory.Exists(ProfrawsDir)
            ? Directory.GetFiles(ProfrawsDir, "*.profraw")
            : [];
        if (profraws.Length == 0)
        {
            Console.WriteLine("  (no .profraw produced - skipping native LCOV export)");
            return false;
        }
        Console.WriteLine($"  Merging {profraws.Length} profraw files -> {Path.GetFileName(MergedProfdata)}");
        var profrawList = Path.Combine(CoverageDir, "profraws.list");
        File.WriteAllText(profrawList, string.Join('\n', profraws));
        Run(["llvm-profdata", "merge", "-sparse", $"-o={MergedProfdata}", "-f", profrawList]);

        if (binaries.Count == 0)
        {
            Console.WriteLine("  (no instrumented binaries ran - skipping native LCOV export)");
            return false;
        }
        Console.WriteLine($"  Exporting LCOV across {binaries.Count} binaries -> {Path.GetFileName(NativeLcov)}");
        var first = binaries[0];
        var rest = binaries.Skip(1);
        List<string> cmd = ["llvm-cov", "export", "-format=lcov", $"-instr-profile={MergedProfdata}", first];
        cmd.AddRange(rest.Select(p => $"-object={p}"));
        Run(cmd, cwd: BaseDir, stdoutFile: NativeLcov);
        return true;
    }

    // ── Phase 3: managed tests ───────────────────────────────────────────────

    static void RunManagedTests()
    {
        Banner("[3/4] Run managed (C#) tests");
        if (Directory.Exists(ManagedDir)) Directory.Delete(ManagedDir, recursive: true);
        Directory.CreateDirectory(ManagedDir);
        Run([
            "dotnet", "test", Path.Combine(BaseDir, "KernelEngine.slnx"),
            "--collect:XPlat Code Coverage",
            "--results-directory", ManagedDir,
            "--nologo", "-v", "m",
            "--",
            "DataCollectionRunSettings.DataCollectors.DataCollector.Configuration.Format=cobertura",
        ], check: false);
    }

    // ── Phase 4: unified report ──────────────────────────────────────────────

    static void BuildReport()
    {
        Banner("[4/4] Generate unified report");
        if (Directory.Exists(ReportDir)) Directory.Delete(ReportDir, recursive: true);
        Directory.CreateDirectory(ReportDir);

        var inputs = new List<string>();
        if (File.Exists(NativeLcov) && new FileInfo(NativeLcov).Length > 0)
            inputs.Add(NativeLcov);
        if (Directory.Exists(ManagedDir))
            inputs.AddRange(Directory.GetFiles(ManagedDir, "*.xml", SearchOption.AllDirectories));

        if (inputs.Count == 0)
        {
            Console.WriteLine("  (no coverage data - nothing to report)");
            return;
        }
        Run([
            "reportgenerator",
            "-reports:" + string.Join(';', inputs),
            "-targetdir:" + ReportDir,
            "-reporttypes:Cobertura;Html",
            "-sourcedirs:" + BaseDir,
            // *tests* is no longer excluded: the GTest suites' own .cpp files are
            // now the primary native coverage signal (see file header).
            "-filefilters:-*Generated*;-*examples*",
            "-classfilters:-*NativeMethods*;-*NativeAnnotation*;-*NativeTypeName*",
        ], check: false);
    }

    // ── Structured summary ───────────────────────────────────────────────────

    class Counts
    {
        public int Covered, Total;
        public double Pct => Total != 0 ? 100.0 * Covered / Total : 0.0;
    }

    static string ModuleFor(string layer, string path)
    {
        var parts = Norm(path).Split('/');
        return layer switch
        {
            // tests/c/kernel/... -> tests/c/kernel ; src/c/<domain>/... -> src/c/<domain>
            "L2" => parts.Length >= 3 ? string.Join('/', parts[..3]) : string.Join('/', parts),
            // src/csharp/<project>/... -> <project>
            "L4" => parts.Length >= 3 ? parts[2] : "(unknown)",
            _ => "(unknown)",
        };
    }

    static string Short(string module) => module.Split('/') is { Length: > 0 } parts ? parts[^1] : module;

    static void PrintSummary()
    {
        var cobertura = Path.Combine(ReportDir, "Cobertura.xml");
        if (!File.Exists(cobertura))
        {
            Console.WriteLine("\n  (no Cobertura.xml - run the pipeline first)\n");
            return;
        }

        var root = XDocument.Load(cobertura).Root!;

        var perLayer = Layers.ToDictionary(l => l.Id, _ => new
        {
            Totals = new Counts(),
            Modules = new Dictionary<string, Counts>(),
        });
        var native = new Counts();
        var managed = new Counts();

        var baseUri = Norm(BaseDir).TrimEnd('/') + "/";

        foreach (var clazz in root.Descendants("class"))
        {
            var filePath = Norm(clazz.Attribute("filename")?.Value ?? "");
            if (filePath.StartsWith(baseUri)) filePath = filePath[baseUri.Length..];

            var lines = clazz.Descendants("line").ToList();
            if (lines.Count == 0) continue;
            var covered = lines.Count(ln => int.Parse(ln.Attribute("hits")?.Value ?? "0") > 0);
            var lineTotal = lines.Count;

            foreach (var (id, _, predicate) in Layers)
            {
                if (!predicate(filePath)) continue;
                var l = perLayer[id];
                l.Totals.Covered += covered;
                l.Totals.Total += lineTotal;
                var module = ModuleFor(id, filePath);
                if (!l.Modules.TryGetValue(module, out var m)) l.Modules[module] = m = new Counts();
                m.Covered += covered;
                m.Total += lineTotal;
                var bucket = id == "L2" ? native : managed;
                bucket.Covered += covered;
                bucket.Total += lineTotal;
                break;
            }
        }

        var total = new Counts { Covered = native.Covered + managed.Covered, Total = native.Total + managed.Total };

        var outLines = new List<string>();
        void W(string s = "") => outLines.Add(s);

        var now = DateTime.Now.ToString("yyyy-MM-dd HH:mm");
        W(new string('=', 70));
        W($"  KernelEngine Coverage - {now}");
        W(new string('=', 70));
        W();
        W($"  TOTAL                                          {total.Pct,5:F1}%   {total.Covered}/{total.Total}");
        W();
        W($"  Native  (C/C++)                                {native.Pct,5:F1}%   {native.Covered}/{native.Total}");
        W($"  Managed (C#)                                   {managed.Pct,5:F1}%   {managed.Covered}/{managed.Total}");
        W();
        W("  Per layer:");
        foreach (var (id, name, _) in Layers)
        {
            var l = perLayer[id];
            W($"    {id} - {name,-22} {l.Totals.Pct,5:F1}%   {l.Totals.Covered}/{l.Totals.Total}");
            foreach (var (module, m) in l.Modules.OrderBy(kv => kv.Key, StringComparer.Ordinal))
                W($"        {Short(module),-38} {m.Pct,5:F1}%");
        }
        W();

        var gaps = new List<(string Module, double Pct, int Total)>();
        foreach (var (id, _, _) in Layers)
            foreach (var (module, m) in perLayer[id].Modules)
                if (m.Total >= GapThresholdLines && m.Pct < GapThresholdPct)
                    gaps.Add((module, m.Pct, m.Total));

        if (gaps.Count > 0)
        {
            W($"  Critical gaps (< {GapThresholdPct:F0}% AND > {GapThresholdLines} lines):");
            foreach (var (module, p, n) in gaps.OrderBy(g => g.Pct))
                W($"    - {Short(module),-38} {p,5:F1}%   ({n} lines)");
            W();
        }
        W($"  HTML report: {Path.Combine(ReportDir, "index.html")}");
        W(new string('=', 70));

        var text = string.Join('\n', outLines);
        File.WriteAllText(SummaryTxt, text);
        Console.WriteLine();
        Console.WriteLine(text);
    }

    // ── Subcommands ──────────────────────────────────────────────────────────

    public static void CmdClean()
    {
        if (Directory.Exists(CoverageDir))
        {
            Console.WriteLine($"Removing {CoverageDir}");
            Directory.Delete(CoverageDir, recursive: true);
        }
        else
        {
            Console.WriteLine($"{CoverageDir} does not exist.");
        }
    }

    public static void CmdRun()
    {
        ConfigureAndBuild();
        var binaries = RunNativeTests();
        MergeNativeProfile(binaries);
        RunManagedTests();
        BuildReport();
        PrintSummary();
    }

    public static void CmdReport() => PrintSummary();
}

static class ScriptPaths
{
    public static string Dir => Impl();
    static string Impl([CallerFilePath] string path = "") => Path.GetDirectoryName(path)!;
}
